### DISCUSSION TITLE
show TOC on mobile / small width screen devices, and JS / CSS refactor talks

### DISCUSSION THREAD

**Discussion by lgh06** [obsidian-html/obsidian-html#262] [Ideas]
my fork version implement TOC on mobile (only on documentation mode) after did tons of CSS / JS tricks, on end of April.  
<video src="https://user-images.githubusercontent.com/1688639/168473186-54bf397a-e871-4e87-bc1a-75d51ccd8e9c.mp4"></video>

it modified one container's flex order and did some media queries, and modified some JS, and modified headings link anchor without `!` (because I do not need tabs mode so I do not need those `#!xx` tricks).
On this process, I found some `onload=func()` `onclick=func()` inside HTML codes,  
maybe we can refactor those to JS codes by `document.querySelector(selector).addEventListener('load or click or others', afunc)`  
also, CSS / CSS class names / CSS ids are hard to distinct which page /area uses which lines of CSS.  

However my fork version became too far away from obsidian-html newest version.  

My question are: 
Is it possible to show TOC on mobile devices, for every modes?  
Is there a better way to implement TOC for mobile on documentation mode?
Do you have any plan to refactor codes like JS event binding / CSS  class names and ids ,  any plan to find a better way to arrage JS codes  ?

thanks.

**Comment by lgh06** [comment]
Addition :  I mainly write JS / Node.js / CSS / frontend things for years, however not so familiar with python.
maybe we could find a better way to arrage this repo's css / js codes, and make it readable / understandable / maintainable for future.

**Comment by lgh06** [comment]
## Some JS refactor advice:  

we can limit some JS functions only have effect on some limited pages, like using
```javascript
if(window.location.href.match(/RegExp/)){
  let doSomethingFunction = () => {};
  document.querySelector('.someClass').addEventListener('click', doSomethingFunction);
}
```

we can also prevent global variable name pollution by wrap all JS codes to a function, like: 
```
(function(){
  // someVar is not on window and won't mass up other things
  let someVar = "";
})();
```

we can also modify onload to a jquery `$.ready()` like function, because ready is much earlier than load event, so our pages can render faster:  

```javascript

(function(){

  function ready(fn) {
    if (document.readyState != 'loading'){
      fn();
    } else {
      document.addEventListener('DOMContentLoaded', fn);
    }
  }
  function someOnLoadFunction(){
    // move existing LoadPage() functions inside here
  };
  ready(someOnLoadFunction);
})();
```

## Also, we can use a CDN to distribute some big JS / CSS files  
There is no need to copy css / js from another repo to this repo.  
https://cdn.jsdelivr.net/gh/obsidian-html/md_mermaid/mermaid.min.js  is same but faster with a local version mermaid.min.js   
I recommend jsdelivr personally because it transforms any github repo files on a CDN, see its home page https://www.jsdelivr.com/?docs=gh , and it also supports grab specific version file, by adding `@version/@tag/@commithash` , like https://cdn.jsdelivr.net/gh/obsidian-html/md_mermaid@master/mermaid.css or https://cdn.jsdelivr.net/gh/obsidian-html/md_mermaid@22c4d60/mermaid.css  , and jsdelivr is stable and work fine worldwide.

we can support configs in yaml, for different CDN providers, in future.

**Comment by dwrolvink** [comment]
# TOC on mobile
If you set the following settings, then the toc is created inline below the first h1 header. (Unless already present on the page).

``` yaml
  features:
    styling: 
      add_toc: True    # add "[TOC]" (Table of Contents) when missing
      toc_pane: False   # removes table of contents from the note and puts it in the right pane (not supported for layout:tabs)
```
This results in an inline TOC like this:
![image](https://user-images.githubusercontent.com/30291690/168483439-effca304-6fde-43a2-8187-7f36d9664eba.png)

At page load the function `obsidian_core_footer.js/LoadTableOfContents()` is run. When `toc_pane: True` this function loads the contents of the toc div, puts them in the side pane. If it is false, it will show the toc div (which is hidden by default by the css).

If we want to show the inline toc when the sidepanel toc is hidden, then adding this css will do the trick:
``` css
@media (max-width: 840px) {
    .toc {
        display: block;
    }
```

Also, we need to reenable the toc_pane of course:
``` yaml
  features:
    styling: 
      add_toc: True    # add "[TOC]" (Table of Contents) when missing
      toc_pane: True   # removes table of contents from the note and puts it in the right pane (not supported for layout:tabs)
```
That will give this behavior:

![image](https://user-images.githubusercontent.com/30291690/168483826-d470dd9d-6827-434f-8ad3-216dc3bacd80.png)
![image](https://user-images.githubusercontent.com/30291690/168483839-fc839a41-d2ba-443b-a4bd-1611c5f45efd.png)
![image](https://user-images.githubusercontent.com/30291690/168483849-2717106b-6161-4a22-8199-c5fee07d7896.png)

Let me know if this is what you meant. I think this is a good default setting, I'll add it regardless.

> Is it possible to show TOC on mobile devices, for every modes?

Adding a TOC inline is available for every mode, so the fix above should work for every mode

> Is there a better way to implement TOC for mobile on documentation mode?

Let me know what you think of my solution :)

> Do you have any plan to refactor codes like JS event binding / CSS class names and ids , any plan to find a better way to arrage JS codes ?

Yes, I'll be the first to say that it is all quite a mess on the frontend side. I've already restructured the backend code many times, and that is pretty well organized now, but I've only just started with the frontend code. The first work was done going from 2.2.0 to 2.3.0. A lot of duplicate css has been removed, templates and different layouts now are more similar and share more code, and all the core js code has been put in the two core files, where the rest have been split off. This should give some more structure.

But any tips / proposals for renaming CSS class names / function calls / event bindings is welcome.

# Refactors
> On this process, I found some onload=func() onclick=func() inside HTML codes, maybe we can refactor those to JS codes by document.querySelector(selector).addEventListener('load or click or others', afunc)

What would the benefit of this be? I kind of like looking at the html code and seeing what onclicks are active on which elements.

The only onload is on the body tag, I guess we can change that to:
``` js
document.addEventListener('DOMContentLoaded', load_page); 
```

> we can limit some JS functions only have effect on some limited pages, like using

useful pattern, thanks, that will come in handy

> we can also prevent global variable name pollution by wrap all JS codes to a function, like

Everything that is global was intended to be global, we will need to look at a var by var basis what can be removed from the global scope and why that would be necessary.

> we can also modify onload to a jquery $.ready() like function, because ready is much earlier than load event, so our pages can render faster

So if I change this code:
``` js
// Onloads
// ----------------------------------------------------------------------------
document.addEventListener('DOMContentLoaded', load_theme);  //set theme as quickly as possible to avoid flickering
document.addEventListener('DOMContentLoaded', load_page);   // all the code that needs to be run when everything is loaded
```

to:
``` js 
(function(){
  function ready(fn) {
    if (document.readyState != 'loading'){
      fn();
    } else {
      document.addEventListener('DOMContentLoaded', fn);
    }
  }
  function someOnLoadFunction(){
    load_theme();
    load_page();
  };
  ready(someOnLoadFunction);
})();
```

Then it should be faster? :thinking: 

I've tried this out with some very basic performance testing, and this solution actually seems to be a bit slower. (On Firefox at least).

**Reply by lgh06** [reply]
thx. i will try inline TOC.

**Comment by dwrolvink** [comment]
The toc code has been added to the master branch https://github.com/obsidian-html/obsidian-html/pull/263/commits/4db1b407fa0bf7a826dea22387f65d8a3ee2fd2c

**Comment by dwrolvink** [comment]
I now notice that in your solution the toc remains available at the top. With my solution if you scroll down you lose the toc. 

You seem to know your way around the code pretty well. If you duplicate the template of the layout that you want, then you can change it to suit your own needs and use that to build your website. That will be a way more stable way to build on top of obsidianhtml, because the template/css is in constant flux. 

If you make your own template then you can do things like this:
``` html
<body>
<div id="header"></div>
<div id="toc-container"></div>
<div id="content-container">{content}</div>
<script>
function populateToc(){
  document.getElementById('toc-container').innerHtml = document.getElementById('toc').innerHtml
}
document.addEventListener('DOMContentLoaded', populateToc);
</body>
```

Which will be way easier than trying to pry it out of the `.container` div.

**Comment by lgh06** [comment]
we shouldn't trust DOMContentLoaded because this event only fire once and our js codes may execute after DOMContentLoaded event fired, that's why we should also check document.readyState

the reason why ready and load have similar performance is those third party big JS / CSS dependencies blocks page render.
Modify load to ready should have some big improvements if we mark some header third party JS as async or defer, however we should write more js logics to fire those functions who need a third party dependency , manually, on one dependency fully loads.

more js codes and more controls like above is common for our projects becoming bigger and more complex, inject JS events inside HTML may not match our advanced requirements in the near future, and html injected JS functions name should exists on global window, that may pollute other vars and may have potential conflict issues .

try to find a better way to arrange JS codes can also make we figure out which page area has which js functions executed easier. inject js to HTML are intuitively and suitable for now but lack advanced control and pollute global vars.
